File: app/lib/share.ts
import { getTier, type TierKey } from "./tier";
import { type Dictionary, type Locale, t } from "./i18n";
import { CLASSIC, type ModeConfig, type ModeId } from "./mode";

export type ShareInput = {
  puzzleNumber: number;
  moves: number;
  // The puzzle's exact minimum number of swaps. Required for tier
  // computation under the ratio-based system.
  minSwaps: number;
  streak: number;
  bonus?: boolean;
  revealed?: boolean;
  locale: Locale;
  // Locale-aware copy: pass the player's current dictionary so the headline,
  // streak line, and bonus tag are in the language they're playing in.
  dict: Dictionary;
  mode?: ModeConfig;
};

// Compact representation of a share slug. Used both in the share URL
// (so the recipient lands on a page with a per-solve OG card) and parsed
// back in app/page.tsx generateMetadata to populate the card.
export type ShareSlug = {
  num: number;
  moves: number | null;
  bonus: boolean;
  revealed: boolean;
  mode?: ModeId;
};

// Tier-specific emoji used in the share headline. Keyed by locale-independent
// tier key so emoji stay constant across languages.
const TIER_EMOJI: Record<TierKey, string> = {
  legendary: "🏆",
  genius: "🧠",
  wordsmith: "📖",
  persistent: "🧩",
  tenacious: "💪",
};

const SOLVED_TILE = "🟩";
const BONUS_TILE = "🟧";
const REVEALED_TILE = "⬜";

// Render an N×N emoji grid encoding the player's result. Solved rows
// fill green; on a bonus solve the corner tiles flip to orange as a
// visual flourish ("rows + columns"); on a reveal the grid is empty.
function buildGrid({
  revealed,
  bonus,
  N,
}: {
  revealed: boolean;
  bonus: boolean;
  N: number;
}): string {
  if (revealed) {
    return Array.from({ length: N }, () => REVEALED_TILE.repeat(N)).join("\n");
  }
  if (!bonus) {
    return Array.from({ length: N }, () => SOLVED_TILE.repeat(N)).join("\n");
  }
  // Bonus solve: corners become orange to flag the columns-also-spell
  // achievement. Reads as a "framed" grid.
  const last = N - 1;
  const corners = new Set([`0,0`, `0,${last}`, `${last},0`, `${last},${last}`]);
  return Array.from({ length: N }, (_, r) =>
    Array.from({ length: N }, (_, c) =>
      corners.has(`${r},${c}`) ? BONUS_TILE : SOLVED_TILE
    ).join("")
  ).join("\n");
}

// Encode a result into a short URL slug. Classic shares stay
// unchanged: "6-12", "6-12-b", "6-r". Hard mode gets an "h" prefix:
// "h6-12", "h6-12-b", "h6-r". The mode is inferred from the URL path
// (/s vs /hard/s) but the slug carries it too so legacy or copy-pasted
// URLs that lose the path still parse correctly.
export function buildShareSlug(s: ShareSlug): string {
  const prefix = s.mode === "hard" ? "h" : "";
  if (s.revealed) return `${prefix}${s.num}-r`;
  const flag = s.bonus ? "-b" : "";
  const moves = s.moves ?? 0;
  return `${prefix}${s.num}-${moves}${flag}`;
}

// Inverse of buildShareSlug. Returns null on anything malformed so the
// page can fall back to default metadata rather than render a broken
// card. Accepts an optional leading "h" for hard mode.
export function parseShareSlug(slug: string): ShareSlug | null {
  const hardMatch = /^h(\d.*)$/.exec(slug);
  const isHard = hardMatch !== null;
  const body = isHard ? hardMatch[1] : slug;
  if (!/^\d+(?:-(?:r|\d+(?:-b)?))?$/.test(body)) return null;
  const parts = body.split("-");
  const num = parseInt(parts[0], 10);
  if (!Number.isFinite(num) || num <= 0) return null;
  const mode: ModeId | undefined = isHard ? "hard" : undefined;
  if (parts.length === 1) {
    return { num, moves: null, bonus: false, revealed: false, mode };
  }
  if (parts[1] === "r") {
    return { num, moves: null, bonus: false, revealed: true, mode };
  }
  const moves = parseInt(parts[1], 10);
  if (!Number.isFinite(moves) || moves < 0) return null;
  const bonus = parts[2] === "b";
  return { num, moves, bonus, revealed: false, mode };
}

// Returns the share split into a `text` half (headline, meta, grid) and a
// separate `url`. Web Share API targets disagree on which field they read:
// WhatsApp / X / iMessage concatenate text+url; Facebook drops `text` and
// only reads `url`, relying on OG tags to unfurl. Passing both fields
// makes every target work. `full` is the joined version used for the
// clipboard fallback.
export function buildSharePayload(input: ShareInput): {
  text: string;
  url: string;
  full: string;
} {
  const {
    puzzleNumber,
    moves,
    minSwaps,
    streak,
    bonus = false,
    revealed = false,
    locale,
    dict,
    mode = CLASSIC,
  } = input;
  const tier = getTier(moves, minSwaps);
  const tierName = t(dict, `tiers.${tier.key}`);
  const tierEmoji = TIER_EMOJI[tier.key] ?? "";
  const swapWord = t(dict, moves === 1 ? "game.swapSingular" : "game.swapPlural");

  const headlineKey = mode.id === "hard" ? "share.headlineSolvedHard" : "share.headlineSolved";
  const revealedKey = mode.id === "hard" ? "share.headlineRevealedHard" : "share.headlineRevealed";
  const headline = revealed
    ? t(dict, revealedKey, { num: puzzleNumber })
    : t(dict, headlineKey, {
        num: puzzleNumber,
        moves,
        swapWord,
        tierEmoji,
        tier: tierName,
      }).trim();

  const meta: string[] = [];
  if (!revealed && streak > 1) meta.push(t(dict, "share.streakMeta", { streak }));
  if (!revealed && bonus) meta.push(t(dict, "share.bonusMeta"));

  const slug = buildShareSlug({
    num: puzzleNumber,
    moves,
    bonus,
    revealed,
    mode: mode.id,
  });
  const localePrefix = locale === "en" ? "" : `/${locale}`;
  const sharePath = mode.id === "hard" ? "/hard/s" : "/s";
  const url = `https://tesserapuzzle.com${localePrefix}${sharePath}/${slug}`;

  const lines = [headline];
  if (meta.length) lines.push(meta.join(" · "));
  lines.push("");
  lines.push(buildGrid({ revealed, bonus, N: mode.N }));
  const text = lines.join("\n");
  return { text, url, full: `${text}\n\n${url}` };
}

File: mobile/lib/src/share.dart
// Port of app/lib/share.ts — share slug encode/decode, the emoji result
// grid (default + colour-blind), and the full share payload (headline +
// meta + grid + CTA). Fixture-locked against test/fixtures/parity.json.
//
// The colour-blind grid swaps the solved tile only (green -> blue); the
// bonus/orange and revealed/white tiles are unchanged (spec §17.2).

import 'i18n.dart';
import 'tier.dart';

const String solvedTile = '\u{1F7E9}'; // 🟩
const String colourBlindSolvedTile = '\u{1F7E6}'; // 🟦
const String bonusTile = '\u{1F7E7}'; // 🟧
const String revealedTile = '\u{2B1C}'; // ⬜

const Map<TierKey, String> tierEmoji = {
  TierKey.legendary: '\u{1F3C6}', // 🏆
  TierKey.genius: '\u{1F9E0}', // 🧠
  TierKey.wordsmith: '\u{1F4D6}', // 📖
  TierKey.persistent: '\u{1F9E9}', // 🧩
  TierKey.tenacious: '\u{1F4AA}', // 💪
};

class ShareSlug {
  const ShareSlug({
    required this.num,
    required this.moves,
    required this.bonus,
    required this.revealed,
    this.mode,
  });

  final int num;
  final int? moves;
  final bool bonus;
  final bool revealed;
  final String? mode; // "hard" or null

  @override
  bool operator ==(Object other) =>
      other is ShareSlug &&
      other.num == num &&
      other.moves == moves &&
      other.bonus == bonus &&
      other.revealed == revealed &&
      other.mode == mode;

  @override
  int get hashCode => Object.hash(num, moves, bonus, revealed, mode);
}

/// Encode a result into a short URL slug: "6-12", "6-12-b", "6-r";
/// hard mode gets an "h" prefix.
String buildShareSlug(ShareSlug s) {
  final prefix = s.mode == 'hard' ? 'h' : '';
  if (s.revealed) return '$prefix${s.num}-r';
  final flag = s.bonus ? '-b' : '';
  final moves = s.moves ?? 0;
  return '$prefix${s.num}-$moves$flag';
}

final _hardSlug = RegExp(r'^h(\d.*)$');
final _slugBody = RegExp(r'^\d+(?:-(?:r|\d+(?:-b)?))?$');

/// Inverse of [buildShareSlug]. Returns null on anything malformed.
ShareSlug? parseShareSlug(String slug) {
  final hardMatch = _hardSlug.firstMatch(slug);
  final isHard = hardMatch != null;
  final body = isHard ? hardMatch.group(1)! : slug;
  if (!_slugBody.hasMatch(body)) return null;
  final parts = body.split('-');
  final num = int.tryParse(parts[0]);
  if (num == null || num <= 0) return null;
  final mode = isHard ? 'hard' : null;
  if (parts.length == 1) {
    return ShareSlug(num: num, moves: null, bonus: false, revealed: false, mode: mode);
  }
  if (parts[1] == 'r') {
    return ShareSlug(num: num, moves: null, bonus: false, revealed: true, mode: mode);
  }
  final moves = int.tryParse(parts[1]);
  if (moves == null || moves < 0) return null;
  final bonus = parts.length > 2 && parts[2] == 'b';
  return ShareSlug(num: num, moves: moves, bonus: bonus, revealed: false, mode: mode);
}

/// N x N emoji grid encoding the result. Solved rows fill green; a bonus
/// solve flips the four corners to orange; a reveal is all-white.
String buildGrid({required bool revealed, required bool bonus, required int n}) {
  if (revealed) {
    return List.generate(n, (_) => revealedTile * n).join('\n');
  }
  if (!bonus) {
    return List.generate(n, (_) => solvedTile * n).join('\n');
  }
  final last = n - 1;
  bool isCorner(int r, int c) => (r == 0 || r == last) && (c == 0 || c == last);
  return List.generate(
    n,
    (r) => List.generate(n, (c) => isCorner(r, c) ? bonusTile : solvedTile).join(),
  ).join('\n');
}

/// The colour-blind text grid: green -> blue only (spec §17.2).
String toColourBlindGrid(String grid) => grid.replaceAll(solvedTile, colourBlindSolvedTile);

class ShareInput {
  const ShareInput({
    required this.puzzleNumber,
    required this.moves,
    required this.minSwaps,
    required this.streak,
    required this.locale,
    required this.dict,
    this.bonus = false,
    this.revealed = false,
    this.mode = 'classic',
  });

  final int puzzleNumber;
  final int moves;
  final int minSwaps;
  final int streak;
  final String locale;
  final Map<String, dynamic> dict;
  final bool bonus;
  final bool revealed;
  final String mode; // "classic" | "hard"
}

class SharePayload {
  const SharePayload({required this.text, required this.url, required this.full});
  final String text;
  final String url;
  final String full;
}

/// Full share payload: headline + optional meta + blank + grid, plus a CTA
/// on solved shares. `text` and `url` are separate because Web Share targets
/// disagree on which they read; `full` is the joined clipboard fallback.
SharePayload buildSharePayload(ShareInput input) {
  final isHard = input.mode == 'hard';
  final n = isHard ? 5 : 4;
  final tier = getTier(input.moves, input.minSwaps);
  final tierName = t(input.dict, 'tiers.${tier.key.name}');
  final emoji = tierEmoji[tier.key] ?? '';
  final swapWord =
      t(input.dict, input.moves == 1 ? 'game.swapSingular' : 'game.swapPlural');

  final headlineKey = isHard ? 'share.headlineSolvedHard' : 'share.headlineSolved';
  final revealedKey =
      isHard ? 'share.headlineRevealedHard' : 'share.headlineRevealed';
  final headline = input.revealed
      ? t(input.dict, revealedKey, {'num': input.puzzleNumber})
      : t(input.dict, headlineKey, {
          'num': input.puzzleNumber,
          'moves': input.moves,
          'swapWord': swapWord,
          'tierEmoji': emoji,
          'tier': tierName,
        }).trim();

  final meta = <String>[];
  if (!input.revealed && input.streak > 1) {
    meta.add(t(input.dict, 'share.streakMeta', {'streak': input.streak}));
  }
  if (!input.revealed && input.bonus) {
    meta.add(t(input.dict, 'share.bonusMeta'));
  }

  final slug = buildShareSlug(ShareSlug(
    num: input.puzzleNumber,
    moves: input.moves,
    bonus: input.bonus,
    revealed: input.revealed,
    mode: input.mode,
  ));
  final localePrefix = input.locale == 'en' ? '' : '/${input.locale}';
  final sharePath = isHard ? '/hard/s' : '/s';
  final url = 'https://tesserapuzzle.com$localePrefix$sharePath/$slug';

  final lines = <String>[headline];
  if (meta.isNotEmpty) lines.add(meta.join(' · ')); // " · "
  lines.add('');
  lines.add(buildGrid(revealed: input.revealed, bonus: input.bonus, n: n));
  if (!input.revealed) {
    lines.add('');
    lines.add(t(input.dict, 'share.challenge'));
  }
  final text = lines.join('\n');
  return SharePayload(text: text, url: url, full: '$text\n\n$url');
}
